src/library/kernels/SpGEMM_EM_kernels.cl

```diff
@@ -0,0 +1,1201 @@
+R"(
+/* ************************************************************************
+* The MIT License (MIT)
+* Copyright 2014-2015 weifengliu
+*  Permission is hereby granted, free of charge, to any person obtaining a copy
+*  of this software and associated documentation files (the "Software"), to deal
+*  in the Software without restriction, including without limitation the rights
+*  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+*  copies of the Software, and to permit persons to whom the Software is
+*  furnished to do so, subject to the following conditions:
+ 
+*  The above copyright notice and this permission notice shall be included in
+*  all copies or substantial portions of the Software.
+ 
+*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+*  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+*  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+*  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+*  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+*  THE SOFTWARE.
+* ************************************************************************ */
+/* ************************************************************************
+*  < A CUDA/OpenCL General Sparse Matrix-Matrix Multiplication Program >
+*  
+*  < See papers:
+*  1. Weifeng Liu and Brian Vinter, "A Framework for General Sparse 
+*      Matrix-Matrix Multiplication on GPUs and Heterogeneous 
+*      Processors," Journal of Parallel and Distributed Computing, 2015.
+*  2. Weifeng Liu and Brian Vinter, "An Efficient GPU General Sparse
+*      Matrix-Matrix Multiplication for Irregular Data," Parallel and
+*      Distributed Processing Symposium, 2014 IEEE 28th International
+*      (IPDPS '14), pp.370-381, 19-23 May 2014.
+*  for details. >
+* ************************************************************************ */ 
+
+#pragma OPENCL EXTENSION cl_khr_byte_addressable_store : enable
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
```

Comment:
The code I've used to properly guard those pragmas in another kernel is:
```
#if defined(DOUBLE) && __OPENCL_VERSION__ < CL_VERSION_1_2
  #ifdef cl_khr_fp64
    #pragma OPENCL EXTENSION cl_khr_fp64 : enable
  #elif defined(cl_amd_fp64)
    #pragma OPENCL EXTENSION cl_amd_fp64 : enable
  #else
    #error "Double precision floating point not supported by OpenCL implementation."
  #endif
#endif
```

The host-side setup code then needs to pass in `-DDOUBLE` when building the kernel for double precision. If the kernel is only single precision, all of the fp64 pragmas are ignored.

This does not fully fix the problem; there still needs to be checks in the host-side code to ensure that the target device supports double precision before launching an fp64 kernel. Even in OpenCL >= 1.2, fp64 support is only an optional core feature.

You'll probably want to do a similar check against `__OPENCL_VERSION__ < CL_VERSION_1_1` for the cl_khr_byte_addressable_store pragma. This was made into a core feature in OpenCL 1.1. We should probably also check for this feature on the host side and cleanly fail when trying to run against a device that only supports OpenCL 1.0 and does not have this feature.